cupyx/fallback_mode/fallback.py

```diff
@@ -141,16 +141,36 @@ def __new__(cls, *args, **kwargs):
             return object.__new__(cls)
 
         cupy_ndarray_init = cp.ndarray(*args, **kwargs)
-        return cls(_stored=cupy_ndarray_init)
+        return cls(_stored=cupy_ndarray_init, _class=cp.ndarray)
 
     def __init__(self, *args, **kwargs):
+
+        _class = kwargs.pop('_class', None)
         _stored = kwargs.pop('_stored', None)
-        if _stored is not None:
-            self._array = _stored
+        if _stored is None:
+            return
+
+        self._cupy_array = None
+        self._numpy_array = None
+        self._class = _class
+
+        if _class is cp.ndarray:
+            self._cupy_array = _stored
+            self._latest = 'cupy'
```

Comment:
We've `np.recarray`, `np.ma.MaskedArray`, `np.matrix`, `np.chararray` which will always be `_numpy_array`. Then, `np.ndarray` can be `_cupy_array` (`cp.ndarray` after data transfer) or `_numpy_array`(no immediate data transfer) if it has CuPy supported `dtype`. Otherwise, `np.ndarray` will remain as it is stored as `_numpy_array` (without data transfer) for `dtypes` such as `|S1`, `object`, `M`etc.
So, as you suggested to minimise data transfer during consecutive fallback, `_latest` flag will be used to determine if recent changes were done in `_cupy_array` or `_numpy_array`.
```python
a = numpy.array([1, 2, 3]) # _cupy_array is [1, 2, 3], _numpy_array is None, _latest is 'cupy'
b = numpy.numpy_method(a) # data transfer is done, _cupy_array is [1, 2, 3], _numpy_array is [1, 2, 3], _latest is 'numpy'
c = numpy.another_numpy_method(a)
# _latest tells us that there is no need for data transfer, a._numpy_array is already most
# recent (but don't be sure about _cupy_array because another_numpy_method may have
# updated a inplace).
# So, when cupy_method will be executed _cupy_array will get updated from
# _numpy_array after data transfer.
```
Hope this helps.

Reply:
Already resolved?

Reply:
Yes, this behaviour is now maintained using `ndarray._remember_numpy` flag.